pkg/schema/definitions/handler.go

```diff
@@ -0,0 +1,197 @@
+package definitions
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/steve/pkg/schema/converter"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+var (
+	internalServerErrorCode = validation.ErrorCode{
+		Status: http.StatusInternalServerError,
+		Code:   "InternalServerError",
+	}
+	notRefreshedErrorCode = validation.ErrorCode{
+		Status: http.StatusServiceUnavailable,
+		Code:   "SchemasNotRefreshed",
+	}
+)
+
+// schemaDefinitionHandler is a byID handler for a specific schema, which provides field definitions for all schemas.
+// Does not implement any method allowing a caller to list definitions for all schemas.
+type schemaDefinitionHandler struct {
+	sync.RWMutex
+
+	// lastRefresh is the last time that the handler retrieved models from kubernetes.
+	lastRefresh time.Time
+	// refreshStale is the duration between lastRefresh and the next refresh of models.
+	refreshStale time.Duration
+	// client is the discovery client used to get the groups/resources/fields from kubernetes.
+	client discovery.DiscoveryInterface
+	// models are the cached models from the last response from kubernetes.
+	models *proto.Models
+	// schemaToModel is a map of the schema name to the model for that schema. Can be used to load the
+	// top-level definition for a schema, which can then be processed by the schemaFieldVisitor.
+	schemaToModel map[string]string
+}
+
+// byIDHandler is the Handler method for a request to get the schema definition for a specifc schema. Will use the
+// cached models found during the last refresh as part of this process.
+func (s *schemaDefinitionHandler) byIDHandler(request *types.APIRequest) (types.APIObject, error) {
+	// pseudo-access check, designed to make sure that users have access to the schema for the definition that they
+	// are accessing.
+	requestSchema := request.Schemas.LookupSchema(request.Name)
+	if requestSchema == nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.NotFound, "no such schema")
+	}
+
+	if s.needsRefresh() {
+		err := s.refresh()
```

Comment:
> Do I understand correctly that models is essentially a cache of data from OpenAPI that is refreshed on request?

You are mostly correct - it's refreshed on request at most once every 5 seconds.

> If so, is this approach realistically risking a lot of cache misses with a time to live of 5 seconds and users interacting in the UI with (I expect) much longer average time between clicks?

I don't think so. Keep in mind that the cache is only invalid when a user has installed a new CRD. I expect that this is an infrequent operation (infrequent in the sense of the overall runtime/life of the system, I do expect that most users use this feature), so the cache misses will only occur in a narrow 5 second time frame after a CRD is installed.

> How realistic is the scenario that, when users click around the UI, they get to wait for the refresh of models (of all schemas, not just the one they requested) before getting the schema they want in a majority of cases?

I do think there's some user experience issues with this approach. However, how visible this is will depend on what the UI needs this for and how often they present this information to the end user. Right now the UI just retries for a given schema when encountering this error, which seems to work ok.

Reply:
> the cache is only invalid when a user has installed a new CRD

Can you help me understand where that is checked?

To me, `needsRefresh` reads ([reference](https://github.com/MbolotSuse/steve/blob/d121dfe9885e99d55a3bb7bc236e26ad166b08c7/pkg/schema/definitions/handler.go#L110)):

```go
return s.lastRefresh.Add(s.refreshStale).Before(time.Now())
```

"This `schemaDefinitionHandler`'s cache needs refresh if the last refresh time is older than 5 seconds" 

Immediately after `needsRefresh` returns `refresh` is called, and that in turn queries OpenAPI data, which might take some time due to the volume of data exchanged.

Therefore I expect "by ID" API calls to block for the whole duration of the OpenAPI query at least once every 5 seconds.

I may very well be missing something...

Reply:
Sorry for the misunderstanding - yes, you are correct. Given that this is all in-cluster (i.e. Rancher or the Agent is talking to the API server in the same cluster) I think this isn't too bad. That being said, I agree that this mechanism isn't optimal and am open to any suggestions on how to improve it.

Reply:
This may be worth a brainstorming discussion offline on the best way to cache the results, but here is some info that makes this problem nontrivial.

1. OpenAPI spec changes whenever a CRD or API extension is modified.
2. OpenAPI spec updates are delayed after an undefined amount of time (multiple seconds after a CRD is created and ready to be used)
3. It is common for users to install hundreds of CRDs successively, which causes a large load on the system if each CRD causes a refresh.

IMO, the second characteristic is what makes it hard to confidently cache results.

Reply:
Summary from offline discussions - In a separate PR, we will implement a refresh method that is:
- Debouncable
  - Multiple attempts to refresh within a given time period will only delay a single refresh
  - Refresh should occur 2 seconds after a CRD/APIService is changed (create/update/delete)
  - The length of time after a CRD is changed that results in a refresh should be customizable by env var
- Has some method of correcting a failed refresh
  - A background refresh should occur every 10 minutes by default, even if no CRDs/APIServices have changed
  - The frequency of the refresh should be customizabble by env var
  
  I'll close this thread since we intend on doing this in a separate PR.